src/Lumi/Components2/Text.purs

```diff
@@ -0,0 +1,270 @@
+  ( Text
+  , TextProps, TextType
+  , text
+
+  , TextModifier
+  , body, strong, emphasized, subtext
+  , muted, color
+
+  , ParagraphProps
+  , paragraph_, paragraph
+
+  , Header
+  , HeaderProps
+  , subsectionHeader
+  , sectionHeader
+  , title
+  , mainHeader
+  ) where
+
+import Prelude
+
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Effect.Unsafe (unsafePerformEffect)
+import Lumi.Components (LumiComponent, PropsModifier, PropsModifier', lumiComponent, propsModifier)
+import Lumi.Components.Color (ColorMap)
+import Lumi.Styles (Style, StyleModifier, StyleProperty)
+import Lumi.Styles as S
+import Lumi.Styles.Theme (LumiTheme(..), TextMap, textFontSize, textLineHeight, textMargin, useTheme)
+import React.Basic (JSX, ReactComponent)
+import React.Basic.DOM as R
+import React.Basic.Emotion as E
+import React.Basic.Hooks as Hooks
+
+-- Text
+
+data Text = Text
+
+data TextType
+  = Body
+  | Strong
+  | Emphasis
+  | Subtext
+
+type TextProps =
+  ( component :: Text
+  , type :: Maybe TextType
+  , content :: String
+  )
+
+type TextElement = ReactComponent { children :: Array JSX, className :: String }
+
+-- | Inline piece of text whose `content` prop is a `String`. Elements of this
+-- | component should preferentially be used only where an inline element is
+-- | expected, that is, from within the contents of a block-level element.
+-- |
+-- | For block-level text elements, please use `paragraph`, `paragraph_`, or the
+-- | header components defined in this module.
+text :: LumiComponent TextProps
+text =
+  unsafePerformEffect $ lumiComponent "Text" defaults \props -> Hooks.do
+    theme <- useTheme
+    pure $
+      E.element (maybe R.span' textElement props.type)
+        { children: [ R.text props.content ]
+        , className: props.className
+        , css: defaultTextStyle theme props.type <> props.css theme
+        }
+  where
+    defaults :: Record TextProps
+    defaults =
+      { component: Text
+      , type: Nothing
+      , content: ""
+      }
+
+    defaultTextStyle :: LumiTheme -> Maybe TextType -> Style
+    defaultTextStyle theme ty =
+      S.css
+        { fontSize: maybe S.inherit (px <<< textFontSize theme <<< textTheme) ty
+        , lineHeight: maybe S.inherit (px <<< textLineHeight theme <<< textTheme) ty
+        , whiteSpace: S.str "pre-wrap"
+        , margin: S.str "0"
+        , padding: S.str "0"
+        }
+
+    textElement :: TextType -> TextElement
+    textElement Body = R.span'
+    textElement Strong = R.strong'
+    textElement Emphasis = R.em'
+    textElement Subtext = R.small'
+
+textTheme :: TextType -> (forall a. TextMap a -> a)
+textTheme =
+  case _ of
+    Body -> _.body
+    Strong -> _.body
+    Emphasis -> _.body
+    Subtext -> _.subtext
+
+-- | The `c` type parameter lets us constrain the type of component to which
+-- | a text modifier may be applied: `Text`, `Header` or any.
+type TextModifier c = forall r. PropsModifier (component :: c, type :: Maybe TextType | r)
```

Comment:
This is neat, I like this.

Reply:
Not sure if this is a good idea now, as we can't use, say, `muted` or `color` with `onDesktop`.

Reply:
I still think this is a good idea though, even if it means changing `onDesktop` or using a different pattern there

Reply:
Not sure if we can change `onDesktop` to fit this pattern given how the types work there. This means that `props` is `(component :: c, type :: Maybe TextType)`, but `onDesktop` needs to be polymorphic on `props`, otherwise it can't provide the default values needed.